Milestones example (#949)
open-contracting/standard: docs/guidance/model/milestones.md

# Milestones

Milestones can be included within the planning, tender, contract and contract implementation sections. 

> * The planning milestone block may describe anticipated milestones during the planning stage, such as the preparation 
> of key studies. These anticipated milestones usually set `milestone/type` to 'preProcurement'.

Reviewer: Remove line-break, remove normative keyword, pluralize 'milestones', clarify that the planning milestones block serves two purposes, add examples for clarity:
```suggestion
* The planning milestones block is used for two types of milestones:
  * Key events in the planning stage, for example, the preparation of an environmental impact assessment, the approval to proceed with a project, or the date of a public consultation. 
 * Anticipated milestones during the contract implementation stage, for example, the date by which goods delivery of the goods is required.
```

> * The tender milestone block should be used to describe anticipated milestones during the proposed delivery 
> of a contract, as well as key dates from the process of tendering and awarding the contract.

Reviewer: Remove use of normative key-word, pluralize 'milestones', clarify that the tender milestones block serves two purposes, add examples for clarity:

```suggestion
* The tender milestones block is used to describe two types of milestone:
  * Key dates in the tender and award stages which are not covered by other fields, for example, the date by which procuring entity will respond to enquiries.
  * Anticipated milestones during the contract implementation stage, for example, the date by which goods need to be delivered.
```

> * The contract milestone block should be used for events related to the signing of the contract documents.

Reviewer: Remove normative keyword, pluralize 'milestones', add example for clarity:

```suggestion
* The contract milestones block is used for events related to the signing of the contract, for example, the date of commercial close in a PPP contract.
```

> * The contract implementation milestone block should be used for any event related to delivery of the contract.

Reviewer: Pluralize 'milestones', correct grammar, add example for clarity

```suggestion
* The contract implementation milestones block is used for any events related to the delivery of the contract, for example, the agreed date by which goods will be delivered.
```

> The nature of the milestone is indicated by the [milestone type codelist](../../../schema/codelists#milestone-type).

Reviewer: Add example:

```suggestion
The nature of the milestone is indicated by the [milestone type codelist](../../../schema/codelists#milestone-type), for example, to distinguish between milestones in the planning section which relate to events in the pre-procurement phase and milestones in the planning section which relate to contract implementation.
```

> The lifecycle of the milestone should be modeled with the fields: `dueDate`, `dateMet`, `dateModified` and [status](../../../schema/codelists/#milestone-status).

Reviewer: Remove normative keyword:

```suggestion
The `dueDate`, `dateMet`, `dateModified` and [`status`](../../../schema/codelists/#milestone-status) are used to track the lifecycle of the milestone.
```

# Milestones

Milestones can be included within the planning, tender, contract and contract implementation sections. 

> * The planning milestones block is used for two types of milestones:

Reviewer: Looking at the HTML version of this page, I think it would be easier to read if we use headings rather than bullets, e.g.

> **Planning**
> The planning milestones block is used for two types of milestones:
> * Key events in the planning stage, for example, the preparation of an environmental impact assessment, the approval to proceed with a project, or the date of a public consultation.
> * Anticipated milestones during the contract implementation stage, for example, the date by which goods delivery of the goods is required.
> 
> **Tender**
> The tender milestones block is used to describe two types of milestone...

 * Anticipated milestones during the contract implementation stage, for example, the date by which goods delivery of the goods is required.

* The tender milestones block is used to describe two types of milestone:
  * Key dates in the tender and award stages which are not covered by other fields, for example, the date by which procuring entity will respond to enquiries.
  * Anticipated milestones during the contract implementation stage, for example, the date by which goods need to be delivered.

* The contract milestones block is used for events related to the signing of the contract, for example, the date of commercial close in a PPP contract.

* The contract implementation milestones block is used for any events related to the delivery of the contract, for example, the agreed date by which goods will be delivered.

The nature of the milestone is indicated by the [milestone type codelist](../../../schema/codelists#milestone-type), for example, to distinguish between milestones in the planning section which relate to events in the pre-procurement phase and milestones in the planning section which relate to contract implementation.

> The `dueDate`, `dateMet`, `dateModified` and [`status`](../../../schema/codelists/#milestone-status) are used to track the lifecycle of the milestone.

Reviewer: ```suggestion
The `dueDate`, `dateMet`, `dateModified` and [`status`](../../../schema/codelists/#milestone-status) fields are used to track the lifecycle of the milestone.
```

## Worked examples

> The following worked examples will show how to fill the milestone block depending on its nature.

Reviewer: ```suggestion
The following worked examples show how to use milestones in different scenarios.
```

### Planning and tender milestones

The planning milestone block should be used only to disclose information about the planning process, such as the 
budget approval, key studies etc. If during the planning stage you have information about tender process milestones, then you

> should populated the tender/milestones block instead.

Reviewer: Suggest moving this to the introductory section under the 'Planning milestones' heading, as it is more guidance than worked example.

In the example below:

* The process starts at the planning stage, and the information about when the budget plan is estimated to be ready 
is available. We create a planning release including a milestone with type set to 'preProcurement', `dueDate` set to the estimated date, and `status` set to 'scheduled'. 
The publisher also knows when the tender notice is estimated to take place. This information is part of the **tender** process.  A `milestone` is created in the `tender/milestones` block with type set to 'preProcurement' and the relevant date set in the `dueDate` field. The tender must have a `status` set to 'planned'.
* The process goes forward and enters the tender stage. We publish a tender release 

> with `tender/status` set to 'active',  set the relevant date in the milestone's `dateMet` field, and set its status to 'met'.

Reviewer: I'm not clear what a budget plan being 'ready' means. Is this a real example? If not, it would be clear to have budget approval as the planning milestone.

Copy-edit for clarity:

```suggestion
The example below includes two OCDS releases:

* A planning release with details of a planned procurement, including the date the budget plan is expected to be ready and the date the tender notice is expected to be issued.
* A tender release with the actual date the tender notice was issued.

In the planning release:

* The date the budget plan is expected to be ready is represented using a milestone in `planning/milestones` with `.type` is set to 'preProcurement' because the milestone relates to the planning stage of the contracting process. `.dueDate` is set to the date and `.status` is set to 'scheduled'.
* The date the tender notice is expected to be issued is represented using a milestone in `tender/milestones` because it relates to the tender stage of the contracting process. `.dueDate` is set to the date and `.status` is set to scheduled.

In the tender release:

* The `.dateMet` field in the tender notice milestone is updated with the actual date the notice was issued and `.status` is set to 'met'.
```

> * To explore differences between the planned and actual date of the tender milestone, users can then compare the values of `tender/milestones/dueDate` and `tender/milestones/dateMet` in a single (compiled) release.

Reviewer: Remove bullet-point:

```suggestion
To explore differences between the planned and actual date of the tender milestone, users can then compare the values of `tender/milestones/dueDate` and `tender/milestones/dateMet` in a single (compiled) release.
```

```eval_rst

.. jsoninclude:: ../../examples/milestones/planning-tender-milestones.json
   :jsonpointer: 
   :expand: releases, planning, milestones, tender, milestones
   :title: planning

.. jsoninclude:: ../../examples/milestones/planning-tender-milestones-2.json
   :jsonpointer: 
   :expand: releases, planning, milestones, tender, milestones
   :title: tender

```

### Contract implementation milestones

#### 1. Project data

> This worked example shows how milestones for project commencement and project completion should be updated in OCDS.

Reviewer: Copy-edit for clarity:

```suggestion
This worked example shows how to use milestones to model the planned and actual start and finish dates for a construction project.
```

> The example includes three updates to a single contracting process (ocds-213czf-000-00001):
> 
> * A contract is awarded to a user committee with milestones for project commencement and project completion.
> * The project starts later than expected. The project commencement milestone is updated with the actual start date and the project completion milestone `status` is set to 'notMet'.
> * The project is completed earlier than expected. The project completion milestone is updated with the actual completion date.
> 
> The OCDS releases for each update are shown below. 

Reviewer: Copy-edit for clarity

```suggestion
The example below includes 3 OCDS releases from a contracting process for the construction of a cycle lane:

* An award release which includes the scheduled start and end dates for the project.
* An implementation update release with the actual start date of the project
* An implementation update release with the actual end date of the project

In the award release:

* The scheduled start and end dates for the project are represented using milestones in `contracts/implementation/milestones` with `.type` set to 'delivery' because they related to the delivery of the contract. The dates are provided in `.dueDate` and `.status` is set to 'scheduled'.
* The publisher has defined their own values for the `.code` field so they can filter and compare start and end date milestones across different contracts.

The first implementation update is published after the project starts but before it completes:

* In the project commencement milestone, `.dateMet` is set to the actual start date and `.status` is set to 'met'. `.dateModified` is set to the date the milestone was updated.
* In the project completion milestone, `.status` is set to 'notMet' since the project is not yet complete and `.dateModified` is set to the date the milestone was updated.

Users can compare the project commencement milestone's `.dueDate` and `.dateMet` fields to determine if the project started on time. Users can also compare the project completion milestone's `.dueDate` and `.dateModified` fields to determine whether the `.status` has been updated since the scheduled completion date.

The second implementation update is published after the project completes:

* In the project completion milestone, `.dateMet` is set to the actual completion date for the project and `.status` is set to 'met'.
```

   :title: implementation-update-1

```

```eval_rst

.. jsoninclude:: ../../examples/milestones/implementation-milestones-3.json
   :jsonpointer: 
   :expand: releases, contracts, implementation, milestones
   :title: implementation-update-2

```

#### 2. Delivery and financing data

This example shows how milestones can be used to keep track of delivery and financing (payment) data in a contracting process.

The procuring entity in the example, the Ministry of Border and Tribal Affairs,
is managing the procurement for the mantainance of a high school building.
Milestones have been set for the delivery and payment of goods and services
related to the project. In the first release shown below, contract information
is released along its implementation milestones.

Once the first milestone has been reached ("Finish the exterior and interior walls"), a new release is created with  the related milestone `status` set to 'met' and the relevant dates added to the `dateMet` and `dateModified` fields.

In a second update, the construction company has received payment for the work done so far, and the milestone for the wall restoration with type 'financing' is updated in a new release. 

> A new `transaction` is also disclosed, with the amount paid to the company.

Reviewer: Please update in-line with the structure I suggested for the first two examples. Thanks!

# Milestones

Milestones can be included within the planning, tender, contract and contract implementation sections. 

* The planning milestone block may describe anticipated milestones during the planning stage, such as the preparation 

> of key studies. Usually using the 'preProcurement' milestone type code.

Reviewer 2: May be clearer to say "These anticipated milestones usually set `milestone/type` to 'preProcurement'."?

of a contract, as well as key dates from the process of tendering and awarding the contract.

* The contract milestone block should be used for events related to the signing of the contract documents.

* The contract implementation milestone block should be used for any event related to delivery of the contract.

The nature of the milestone is indicated by the [milestone type codelist](../../../schema/codelists#milestone-type).

The lifecycle of the milestone should be modeled with the fields: `dueDate`, `dateMet`, `dateModified` and [status](../../../schema/codelists/#milestone-status).

## Worked examples

The following worked examples will show how to fill the milestone block depending on its nature.

### Planning and tender milestones

The planning milestone block should be used only to disclose information about the planning process, such as the 
budget approval, key studies etc. If during the planning stage you have information about tender process milestones, then you
should populated the tender/milestones block instead.

In the example below:

* The process starts at the planning stage, and the information about when the budget plan is estimated to be ready 

> is available. We create a planning release including a `preProcurement` milestone type with the estimated date set 

Reviewer 2: I think for [codelist codes](https://ocds-standard-development-handbook.readthedocs.io/en/latest/meta/style_guide.html#text-formatting), the style guide suggests we use single quotes rather than backticks.

Maybe clearer:
We create a planning release including a `preProcurement` milestone type -> We create a planning release including a `milestone` with `type` set to 'preProcurement', `dueDate` set to the estimated date, and `status` set to 'scheduled'

in the `dueDate` field and `scheduled` status. The publisher also knows when the tender notice is estimated to take place. 
This information is part of the **tender** process. A `preProcurement` milestone type with a `dueDate` is created 

> at the `tender/milestones` block. The tender must have the `planned` status.

Reviewer 2: In the example, the tender milestone is missing a `type` of 'preProcurement'.
Also, is "Procurement Plan" the correct title for the tender milestone, since the text refers to the publisher knowing when the tender notice is estimated to take place"?

Reviewer 2: The tender must have the `planned` status -> The tender must have a `status` set to 'planned' .

Reviewer 2: A `preProcurement` milestone type with a `dueDate` is created at the `tender/milestones` block -> A `milestone`  is created in the 'tender/milestones' block with `type` set to 'preProcurement' and the relevant date set in the `dueDate` field.

Author: @pindec I think that you can have a tender estimated date in the procurement plan. At least in Paraguay that that information is available at the planning stage. But I can change the milestone title to be clearer, which title you suggest?

Reviewer 2: @yolile Ah, thanks, I understand, it's the procurement plan's estimation of a tender notice publication date. 

How about: 

"Tender notice publication estimation from procurement plan".

Does that work in the Paraguay context?

Author: yes, thank you, I will change the title to that!

* The process goes forward and enters the tender stage. We publish a tender release 

> with `tender/status` = `active`, the actual dates for milestones in `tender/milestones/dateMet` and `met` status.

Reviewer 2: `tender/status` = `active` -> `tender/status` set to 'active'

Reviewer 2: the actual dates for milestones in `tender/milestones/dateMet` and `met` status. -> set the relevant date in the `milestone`'s `dateMet` field, and set its `status` to 'met'.

> 

Reviewer 2: Could we also add something in here to support use of the data?

For example, "To explore differences between the planned and actual date of the tender milestone, users can then compare the values of tender/milestones/dueDate and tender/milestones/dateMet in a single (compiled) release."

```eval_rst

.. jsoninclude:: ../../examples/planning-tender-milestones.json
   :jsonpointer: 
   :expand: releases, planning, milestones, tender, milestones
   :title: planning

.. jsoninclude:: ../../examples/planning-tender-milestones-2.json
   :jsonpointer: 
   :expand: releases, planning, milestones, tender, milestones
   :title: tender

```

### Contract implementation milestones

#### 1. Project data

This worked example shows how milestones for project commencement and project completion should be updated in OCDS.

The example includes three updates to a single contracting process (ocds-213czf-000-00001):

* A contract is awarded to a user committee with milestones for project commencement and project completion.

> * The project starts later than expected. The project commencement milestone is updated with the actual start date and the project completion milestone `status` is set to `notMet`.

Reviewer 2: `notMet` -> 'notMet'

   :jsonpointer: 
   :expand: releases, contracts, implementation, milestones
   :title: implementation

```

```eval_rst

.. jsoninclude:: ../../examples/implementation-milestones-2.json
   :jsonpointer: 
   :expand: releases, contracts, implementation, milestones
   :title: implementation-update-1

```

```eval_rst

.. jsoninclude:: ../../examples/implementation-milestones-3.json
   :jsonpointer: 
   :expand: releases, contracts, implementation, milestones
   :title: implementation-update-2

```

#### 2. Delivery and financing data

This example shows how milestones can be used to keep track of delivery and financing (payment) data in a contracting process.

The procuring entity in the example, the Ministry of Border and Tribal Affairs,
is managing the procurement for the mantainance of a high school building.
Milestones have been set for the delivery and payment of goods and services
related to the project. In the first release shown below, contract information
is released along its implementation milestones.

Once the first milestone has been reached ("Finish the exterior and interior walls"), a new release is created with the related

> milestone marked as "met" in the scheduled date.

Reviewer 2: the related milestone marked as "met" -> the related milestone `status` set to 'met'

 "in the scheduled date" -> and the relevant dates added to the `dateMet` and `dateModified` fields.

In a second update, the construction company has received payment for the work done so far, and the milestone related to the

> payment for the walls restoration is updated in a new release. A new transaction is also disclosed, with the amount paid to the 

Reviewer 2: maybe clarify which milestone e.g. 
"the milestone related to the payment for the wall restoration with `type` 'financing' is updated"

transaction -> `transaction`

company.

> ```eval_rst

Reviewer 2: The example milestone description fields should either be filled or omitted rather than have  "description": "."

.. jsoninclude:: ../../examples/af-implementation-milestones-1.json
   :jsonpointer: 
   :expand: releases, contracts, implementation, milestones
   :title: implementation

```

```eval_rst

.. jsoninclude:: ../../examples/af-implementation-milestones-2.json
   :jsonpointer: 
   :expand: releases, contracts, implementation, milestones
   :title: implementation-update-1

```

```eval_rst

.. jsoninclude:: ../../examples/af-implementation-milestones-3.json
   :jsonpointer: 
   :expand: releases, contracts, implementation, milestones, transactions
   :title: implementation-update-2

```

Notes:

> * Contract implementation milestones should be kept updated as things progress.

Reviewer 2: maybe add:
", with the relevant date added to the `dateModified` field"?

* At the point of contract signature, a comparison between `tender/milestones` and `contract/implementation/milestones` 

> with a `milestone/type` of `delivery` or `reporting` should reveal how the negotiated contract differs from what was set out at tender time.

Reviewer 2: `delivery` -> 'delivery'
`reporting` -> 'reporting'

# Milestones

Milestones can be included within the planning, tender, contract and contract implementation sections. 

* The planning milestone block may describe anticipated milestones during the planning stage, such as the preparation 
of key studies. Usually using the 'preProcurement' milestone type code.

* The tender milestone block should be used to describe anticipated milestones during the proposed delivery 
of a contract, as well as key dates from the process of tendering and awarding the contract.

* The contract milestone block should be used for events related to the signing of the contract documents.

* The contract implementation milestone block should be used for any event related to delivery of the contract.

The nature of the milestone is indicated by the [milestone type codelist](../../schema/codelists#milestone-type).

The lifecycle of the milestone should be modeled with the fields: `dueDate`, `dateMet`, `dateModified` and [status](../../schema/codelists/#milestone-status).

## Worked example

> In the example below:

Reviewer 2: maybe "example" -> "examples" to make it clearer that people can switch between the 3 cases? (unless it's intended to be one example!)

And add some explanation on whether 2 and 3 are different possible examples following on from the first (see comments on json files, above).

* A contract is awarded to a user committee with milestones for project commencement and project completion.

> * The project starts later than expected. The project commencement milestone is updated with actual start date and the project completion milestone `status` is set to `notMet`.

Reviewer 2: "with actual start date" -> "with the actual start date" (to be consistent with the example below.

* The project is completed earlier than expected. The project completion milestone is updated with the actual completion date.

```eval_rst

> .. jsoninclude:: ../examples/implementation-milestones-1.json

Reviewer 2: Check whether this needs to be "../../examples" due to new structure.
